show_dev() needs to check that port is non-zero.

Signed-off-by: Jack Morgenstein <jackm@mellanox.co.il>

---
 drivers/infiniband/core/user_mad.c |   30 ++++++++++++++++++++++++------
 1 file changed, 24 insertions(+), 6 deletions(-)

Index: ofed_kernel/drivers/infiniband/core/user_mad.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/core/user_mad.c
+++ ofed_kernel/drivers/infiniband/core/user_mad.c
@@ -939,6 +939,20 @@ static struct ib_client umad_client = {
 	.remove = ib_umad_remove_one
 };
 
+static ssize_t show_dev(struct class_device *class_dev, char *buf)
+{
+	struct ib_umad_port *port = class_get_devdata(class_dev);
+
+	if (!port)
+		return -ENODEV;
+
+	if (class_dev == port->class_dev)
+		return print_dev_t(buf, port->dev->dev);
+	else
+		return print_dev_t(buf, port->sm_dev->dev);
+}
+static CLASS_DEVICE_ATTR(dev, S_IRUGO, show_dev, NULL);
+
 static ssize_t show_ibdev(struct class_device *class_dev, char *buf)
 {
 	struct ib_umad_port *port = class_get_devdata(class_dev);
@@ -994,12 +1008,14 @@ static int ib_umad_init_port(struct ib_d
 	if (cdev_add(port->dev, base_dev + port->dev_num, 1))
 		goto err_cdev;
 
-	port->class_dev = class_device_create(umad_class, NULL, port->dev->dev,
+	port->class_dev = class_device_create(umad_class, port->dev->dev,
 					      device->dma_device,
 					      "umad%d", port->dev_num);
 	if (IS_ERR(port->class_dev))
 		goto err_cdev;
 
+	if (class_device_create_file(port->class_dev, &class_device_attr_dev))
+		goto err_class;
 	if (class_device_create_file(port->class_dev, &class_device_attr_ibdev))
 		goto err_class;
 	if (class_device_create_file(port->class_dev, &class_device_attr_port))
@@ -1014,7 +1030,7 @@ static int ib_umad_init_port(struct ib_d
 	if (cdev_add(port->sm_dev, base_dev + port->dev_num + IB_UMAD_MAX_PORTS, 1))
 		goto err_sm_cdev;
 
-	port->sm_class_dev = class_device_create(umad_class, NULL, port->sm_dev->dev,
+	port->sm_class_dev = class_device_create(umad_class, port->sm_dev->dev,
 						 device->dma_device,
 						 "issm%d", port->dev_num);
 	if (IS_ERR(port->sm_class_dev))
@@ -1023,6 +1039,8 @@ static int ib_umad_init_port(struct ib_d
 	class_set_devdata(port->class_dev,    port);
 	class_set_devdata(port->sm_class_dev, port);
 
+	if (class_device_create_file(port->sm_class_dev, &class_device_attr_dev))
+		goto err_sm_class;
 	if (class_device_create_file(port->sm_class_dev, &class_device_attr_ibdev))
 		goto err_sm_class;
 	if (class_device_create_file(port->sm_class_dev, &class_device_attr_port))
@@ -1035,13 +1053,13 @@ static int ib_umad_init_port(struct ib_d
 	return 0;
 
 err_sm_class:
-	class_device_destroy(umad_class, port->sm_dev->dev);
+	class_device_unregister(port->sm_class_dev);
 
 err_sm_cdev:
 	cdev_del(port->sm_dev);
 
 err_class:
-	class_device_destroy(umad_class, port->dev->dev);
+	class_device_unregister(port->class_dev);
 
 err_cdev:
 	cdev_del(port->dev);
@@ -1059,8 +1077,8 @@ static void ib_umad_kill_port(struct ib_
 	class_set_devdata(port->class_dev,    NULL);
 	class_set_devdata(port->sm_class_dev, NULL);
 
-	class_device_destroy(umad_class, port->dev->dev);
-	class_device_destroy(umad_class, port->sm_dev->dev);
+	class_device_unregister(port->class_dev);
+	class_device_unregister(port->sm_class_dev);
 
 	cdev_del(port->dev);
 	cdev_del(port->sm_dev);
